lecture-notes/3-bld.md

```diff
@@ -418,20 +420,284 @@ order:
+recipe will be used.
+
+```make
+hello.o: hello.c world.h
+	# still going to do $(CC) etc
+```
+
+Note that if you are, for example, adding more `.o` dependencies to a big C
+binary and still planning on using the implicit rule, you must make sure that
+the target matches at least one of the `.o` dependencies or the rule won't
+match:
+
+```make
+hello: hello.o world.o foo.o bar.o
+```
+
+Because `hello` shares a stem with `hello.o`, the rule will match. Also note
+that all of the other `.o` files are not the target of any rule! They use the
+implicit `%.o: %.c` pattern rule.
+
+#### Pitfalls
+
+It's not all fun and rainbows. Implicit rules are convenient, and you should
+use them if you want, but remember the following:
+
+* Header files not automatic prerequisites (but you can add them yourself, as
+  above)
+* Implicit rules are not explicit, which some people and projects value
+
+So they may not be the best for you and your team.
+
 ### Silencing commands
-### ?= and += assignment
+
+Build output can be noisy. When you run a command anywhere in a Makefile, Make
+prints the command (not its output---the command itself) before running it.
+This can be useful for debugging, but sometimes the command's only purpose is
+to print something else (e.g. `echo`). In this case, it's not helpful. To
+prevent the command from being printed, preface its line in your recipe with
+`@`.
+
+Without silencing:
+
+```make
+.PHONY: mytarget
+mytarget:
+	echo "hello"
+```
+
+```console
+$ make mytarget
+echo "hello"
+hello
+$
+```
+
+With silencing:
+
+```make
+$ cat Makefile
+.PHONY: mytarget
+mytarget:
+	@echo "hello"
+```
+
+```console
+$ make mytarget
+hello
+$
+```
+
+### `?=` and `+=` assignment
+
+We already saw `=` (recursively expanded) and `:=` (simply expanded)
+assignments for variables. Now we are going to look at `?=` and `+=`.
+
+**`?=` assignment**:
+
+* Only sets the variable if it hasn't been defined yet
+* Doesn't override variables that have been defined, *even if they're empty*
+* Is always recursively-expanded
+
+**`+=` assignment**:
+
+* Appends the given text to the end of an existing variable
+* If not yet defined, is recursively-expanded; otherwise, is expanded as
+  specified by the previous definition
+
+The `+=` assignment is useful for building up lists of things in a Makefile.
+That's pretty much only what we've seen it used for.
+
+The `?=` assignment is excellent for allowing the person running `make` to
+customize the execution of the Makefile without editing it. Speaking of
+which...
+
 ### Variables, overrides, and the environment
+
+<!-- TODO -->
+
+#### Environment variables
+
+You learned about environment variables in the command-line module. They are a
+great way to pass around snippets of data between programs. As such, they are
+commonly used in Make. Make even *imports all environment variables* as Make
+variables. For example, if you have the following Makefile:
```

Comment:
TODO for @tchebb: remind myself of the nuances here so I can review this section more thoroughly